archey.1
.\" Please, before submitting any change, run:
.\" `groff -man -Tascii -z archey.1`

.TH ARCHEY4 1 "${DATE}" "archey4 ${VERSION}" "Archey4 man page"

.SH NAME
archey4 \- A simple system information tool written in Python

.SH SYNOPSIS
\fBarchey\fR [options]
.br
\fBarchey4\fR [options]

.SH DESCRIPTION
Archey4 is a \fBmaintained\fR fork of the original Archey Linux system
tool.
.br
The original Archey program had been written by Melik Manukyan
in 2009, and quickly abandoned in 2011.
.br
At first, it only supported Arch Linux distribution, further support
had been added afterwards.
.br
Many forks popped in the wild due to inactivity, but this one attends
since 2017 to succeed where the others failed:
.br
Remain \fImaintained\fR, \fIcommunity-driven\fR and
\fIhighly-compatible\fR with yesterday's and today's systems.

.SH OPTIONS
.IP "-h, --help"
show help message and exit

.IP "-c, --config-path PATH"
path to a configuration file, or a directory containing a `config.json`

.IP "-d, --distribution IDENTIFIER"
supported distribution identifier to show the logo of, pass `unknown` to list them

.IP "-j, --json"
output entries data to JSON format, use multiple times to increase
indentation

.IP "-l, --logo-style IDENTIFIER"
alternative logo style identifier to show instead of the distribution default one.
For instance, you can try '\fBretro\fR' to prefer old Apple's logo on Darwin
platforms. Pass '\fBnone\fR' to completely hide distribution logo.

.IP "-s, --screenshot [FILENAME]"
take a screenshot once execution is done, optionally specify a target
path

.IP "-v, --version"
show program's version number and exit

.P
Archey will regularly run in terminal text output mode if no argument
is passed.

.SH ENVIRONMENT VARIABLES
.IP NO_COLOR
prevent ANSI-colored output, see <https://no-color.org/>

.IP CLICOLOR
prefer ANSI-colored output when the program isn't piped, see
<https://bixense.com/clicolors/>

.IP CLICOLOR_FORCE
force ANSI-colored output, see <https://bixense.com/clicolors/>

.IP DO_NOT_TRACK
prevent connections to external services on the Internet, see
<https://consoledonottrack.com/>

.SH EXIT STATUS
Archey exits with \fB0\fR on success and \fB1\fR on failure.
.br
On arguments parsing error, it exits with \fB2\fR.

.SH FILES
.I /etc/archey4/config.json
.br
.I ~/.config/archey4/config.json
.br
.I ./config.json
.PP
Please refer to \fBREADME.md\fR for further documentation about
configuration files.

.SH BUGS
Please report any bug to <https://github.com/HorlogeSkynet/archey4/> or
<mailto:samuel+archey@forestier.app>.

.SH SEE ALSO
NEOFETCH(1), SCREENFETCH(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project (partially) adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- `Model` support for Raspberry Pi 5+
- `none` logo style to completely hide distribution logo
- AppArmor confinement profile (included in Debian and AUR packages)

### Changed
- `Model` honor `/proc/device-tree/model` when it exists

### Removed
- `_distribution` protected attribute from `Output` class

## [v4.14.3.0] - 2024-04-06
### Added
- Official Armbian distribution support
- Official EndeavourOS distribution support
- (Nerd Fonts) icons for entries support
- Display server protocol to `WindowManager`
- New Red Hat logo "The Hat"

### Changed
- Allow `df` output to contain Unicode characters
- Extend logos consistency test case to all logo styles
- `WindowManager` API value format: now an object with `name` and `display_server_protocol` attributes
- Old Red Hat logo is available as alternate logo `shadowman`

## [v4.14.2.0] - 2023-08-26
### Added
- Official Nobara distribution support
- Standalone build procedure using Nuitka
- Darwin "WWDC 2020" (`wwdc`) alternate logo
- Darwin "retro" low intensity (`retro_low`) alternate logo
- New `show_link_local` option to allow link-local IP addresses hiding for `LAN_IP`

### Changed
- Log `uptime` errors at warning level
- Honor `PATH` when looking for `vcgencmd` in `Temperature`
- Fallback to HTTP properly on DNS connection errors for `WAN_IP`

## [v4.14.1.0] - 2023-01-20
### Added
- Python 3.12 official support
- Official Enso OS distribution support
- Official Univalent and Moevalent distributions support
- New `logo_style` config option for preferred logo style

### Changed
- Prefer Ubuntu 20.04 to run Python 3.6 CI job
- Fix `CPU` detection on aarch64 platforms lacking of ACPI PPTT

## [v4.14.0.1] - 2022-10-08
### Changed
- Fix warning related to "broken" `Custom` entry
- Fix `PermissionError` when opening restricted configuration files
- Prefer `macOS-latest` for unit tests in integration jobs matrix

## [v4.14.0.0] - 2022-09-22
### Breaking
- Project internal versioning update (v4.Y.Z -> 4.X.Y.Z)
- Distributions logo preference over Windows' one in WSL environments

### Removed
- `dmidecode` usage in `Model` for virtual environments info gathering

### Added
- `Custom` entry
- `Load Average` entry
- KWin Wayland WM detection
- Windows Terminal detection
- Missing method static types
- `CPU`, `RAM` and `Model` FreeBSD support
- Official Guix System distribution support
- Official Siduction Linux distribution support

### Changed
- Improve Rocky Linux logo
- Fix macOS APFS volumes duplication
- `CPU` and `Model` entries initialization
- Prevent vendor name duplication in `Model`
- Fix WSL virtual environment detection (without systemd) in `Model`
- Project code base formatted using Black and isort

## [v4.13.4] - 2022-03-20
### Added
- Support for `pacstall` package manager
- Official Buildroot distribution support
- New `entries_color` config option to tune entries color
- New `sensors_excluded_subfeatures` temperature config option to exclude specific sub-features

### Changed
- Update fpm to v1.14.1 for .DEB and .RPM packages building
- Improve `sensors_chipsets` temperature config option documentation

## [v4.13.3] - 2021-11-13
### Added
- Official Rocky Linux distribution support
- Official upcoming support for Python 3.11
- Support for `pkgin` (NetBSD) package manager
- Support for motherboard DMI information in `Model`
- Document terminal monospaced font recommendation for ASCII art

### Changed
- Fix possible `Uptime` discrepancies on macOS 10.12+
- Prevent `system_profiler` STDERR logs output on macOS
- Remove square brackets around architecture from `Distro` entry output
- `virt-what` and/or `dmidecode` probing for `Model`, even for unprivileged users

## [v4.13.2] - 2021-10-16
### Added
- Support for `maim` ("Make image") screenshot back-end

### Changed
- Fix `df` output parsing when file-systems column contain white-spaces in `Disk`
- Prevent program crash when calling external programs raise `PermissionError` exception
- Prevent program crash when reading from `/sys/` file system is not allowed

## [v4.13.1] - 2021-09-04
### Changed
- Only run `pkg` on \*BSD systems
- Improve Darwin (Apple) logo design
- Fix text width overlap regression introduced in v4.13.0
- Fix `brew` `Packages` count on Apple Silicon and GNU/Linux
- Prevent entries `disabled` special field from being propagated in `options`

## [v4.13.0] - 2021-08-29
### Added
- Official NetBSD support
- Archey official project logo (CC-BY Brume)
- Protected `_logger` attribute for `Entry` objects
- PEP-561 compliance (Distributing and Packaging Type Information)

### Changed
- Logos are now right-padding "unaware"
- Fix Pop!\_OS logo (reverted) coloration
- Fix `sensors_chipsets` option (when containing multiple values)
- Fix potential configuration file encoding issues on Windows platforms
- Improve logging style of multiple lines `sensors` error messages
- Extend Android system auto-detection (as CPython standard library does)

## [v4.12.0] - 2021-05-08
### Added
- `Kernel` name
- CHANGELOG.md file
- Homebrew formula in the default tap
- `CLICOLOR[_FORCE]` environment variables support
- `one_line` configuration option for `LAN_IP` & `WAN_IP` entries
- `TERM_PROGRAM_VERSION` environment variable support for `Terminal`
- `USER`, `LOGNAME` & `LNAME` environment variables support for `User` entry
- Alternative logo style support (see `-l` option with `retro` parameter for Darwin)

### Changed
- Prevent text truncation when writing output to a pipe
- "Pretty names" by default for multiple-words entries (`WanIP` -> `WAN IP`)
- `distro` & `netifaces` Python dependencies are now "frozen" to allow upstream breakages

## [v4.11.0] - 2021-03-21
### Added

archey/archey.py
#!/usr/bin/env python3

"""
Archey main file.
It loads each entry as a different class coming from the `entries` module.
Logos are stored under the `logos` module.
"""

import os
import sys

from enum import Enum
from subprocess import check_output

from .output import Output
from .configuration import Configuration
from .entries.user import User as e_User
from .entries.hostname import Hostname as e_Hostname
from .entries.model import Model as e_Model
from .entries.distro import Distro as e_Distro
from .entries.kernel import Kernel as e_Kernel
from .entries.uptime import Uptime as e_Uptime
from .entries.window_manager import WindowManager as e_WindowManager
from .entries.desktop_environment import DesktopEnvironment as e_DesktopEnvironment
from .entries.shell import Shell as e_Shell
from .entries.terminal import Terminal as e_Terminal
from .entries.packages import Packages as e_Packages
from .entries.temperature import Temperature as e_Temperature
from .entries.cpu import CPU as e_CPU
from .entries.gpu import GPU as e_GPU
from .entries.ram import RAM as e_RAM
from .entries.disk import Disk as e_Disk
from .entries.lan_ip import LanIp as e_LanIp
from .entries.wan_ip import WanIp as e_WanIp
from .constants import COLOR_DICT

# ---------- Global variables --------- #

# We create a global instance of our `Configuration` Class
CONFIG = Configuration()

# We'll list the running processes only one time
try:
    PROCESSES = check_output(
        ['ps', '-u' + str(os.getuid()) if os.getuid() != 0 else '-ax',
         '-o', 'comm', '--no-headers'], universal_newlines=True
    ).splitlines()

except FileNotFoundError:
    print('Please, install first `procps` on your distribution.',
          file=sys.stderr)
    exit()

# ----------- Classes Index ----------- #

class Classes(Enum):
    """
    One more enumeration to store and declare each one of our entries.
    The string representation of keys will act as entries names.
    Values will be set under the `value` attribute of each class.
    """
    User = {
        'class': e_User,
        'kwargs': {
            'not_detected': CONFIG.get('default_strings')['not_detected']
        }
    }
    Hostname = {'class': e_Hostname}
    Model = {
        'class': e_Model,
        'kwargs': {
            'virtual_environment': CONFIG.get(
                'default_strings'
            )['virtual_environment'],
            'bare_metal_environment': CONFIG.get(
                'default_strings'
            )['bare_metal_environment'],
            'not_detected': CONFIG.get(
                'default_strings'
            )['not_detected'],
        }
    }
    Distro = {'class': e_Distro}
    Kernel = {'class': e_Kernel}
    Uptime = {'class': e_Uptime}
    WindowManager = {
        'class': e_WindowManager,
        'kwargs': {
            'processes': PROCESSES,
            'not_detected': CONFIG.get('default_strings')['not_detected'],
        }
    }
    DesktopEnvironment = {
        'class': e_DesktopEnvironment,
        'kwargs': {
            'processes': PROCESSES,
            'not_detected': CONFIG.get('default_strings')['not_detected'],
        }
    }
    Shell = {
        'class': e_Shell,
        'kwargs': {
            'not_detected': CONFIG.get('default_strings')['not_detected'],
        }
    }
    Terminal = {
        'class': e_Terminal,
        'not_detected': CONFIG.get('default_strings')['not_detected'],
        'use_unicode': CONFIG.get('colors_palette')['use_unicode'],
        'clear_color': COLOR_DICT['clear']
    }
    Packages = {
        'class': e_Packages,
        'kwargs': {
            'not_detected': CONFIG.get('default_strings')['not_detected']
        }
    }
    Temperature = {
        'class': e_Temperature,
        'kwargs': {
            'use_fahrenheit': CONFIG.get('temperature')['use_fahrenheit'],
            'char_before_unit': CONFIG.get('temperature')['char_before_unit'],
            'not_detected': CONFIG.get('default_strings')['not_detected']
        }
    }
    CPU = {'class': e_CPU}
    GPU = {
        'class': e_GPU,
        'kwargs': {
            'not_detected': CONFIG.get('default_strings')['not_detected']
        }
    }
    RAM = {
        'class': e_RAM,
        'kwargs': {
            'sensor_color': COLOR_DICT['sensors'],
            'clear_color': COLOR_DICT['clear']
        }
    }
    Disk = {
        'class': e_Disk,
        'kwargs': {
            'sensor_color': COLOR_DICT['sensors'],
            'clear_color': COLOR_DICT['clear']
        }
    }
    LAN_IP = {
        'class': e_LanIp,
        'kwargs': {
            'ip_max_count': CONFIG.get('ip_settings')['lan_ip_max_count'],
            'no_address': CONFIG.get('default_strings')['no_address']
        }
    }
    WAN_IP = {
        'class': e_WanIp,
        'kwargs': {
            'ipv6_support': CONFIG.get('ip_settings')['wan_ip_v6_support'],
            'ipv6_timeout': CONFIG.get('timeout')['ipv6_detection'],
            'ipv4_timeout': CONFIG.get('timeout')['ipv4_detection'],
            'no_address': CONFIG.get('default_strings')['no_address']
        }
    }


# ---------------- Main --------------- #

def main():
    """Simple entry point"""
    output = Output()
    for key in Classes:
        if CONFIG.get('entries', {}).get(key.name, True):
            kwargs = key.value.get('kwargs', {})
            value = key.value['class'](**kwargs).value
            output.append(key.name, value)

    output.output()


if __name__ == '__main__':
    main()
